### PR TITLE
Create custom types for different kinds of IDs

### PR DESCRIPTION
**Pull request by Benzbeeb**

### CODE REVIEW

File: chain/cmd/bandoracled/main.go [hunk 1]
```diff
@@ -100,7 +100,7 @@ func handleRequest(requestID int64) (sdk.TxResponse, error) {
 
 	chanQueryParallelInfo := make(chan queryParallelInfo, len(request.RawDataRequests))
 	for _, rawRequest := range request.RawDataRequests {
-		go func(externalID, dataSourceID int64, calldata []byte) {
+		go func(externalID , dataSourceID int64, calldata []byte) {
```
Comment (sorawit) [inline chain/cmd/bandoracled/main.go]: autoformat?

File: chain/cmd/bandoracled/main.go [hunk 2]
```diff
@@ -100,7 +100,7 @@ func handleRequest(requestID int64) (sdk.TxResponse, error) {
...
 			info := queryParallelInfo{externalID: externalID, answer: []byte{}, err: nil}
```
Comment (sorawit) [inline chain/cmd/bandoracled/main.go:134]: I think this function can accept custom ID types.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/cmd/bandsv/main.go [hunk 3]
```diff
@@ -109,7 +109,7 @@ func handleRequestData(c *gin.Context) {
 	if reqType == Asynchronous {
 		txr, err := bandClient.SendTransaction(
 			zoracle.NewMsgRequestData(
-				req.OracleScriptID,
+				zoracle.OracleScriptID(req.OracleScriptID),
```
Comment (taobun) [inline chain/cmd/bandsv/main.go]: Fix at request type please

File: chain/x/zoracle/internal/types/querier.go [hunk 4]
```diff
@@ -77,15 +77,15 @@ func NewOracleScriptQuerierInfo(
 }
 
 type RequestQuerierInfo struct {
-	ID              int64                          `json:"id"`
+	ID              RequestID                      `json:"id"`
```
Comment (taobun) [inline chain/x/zoracle/internal/types/querier.go:86]: OracleScriptQuerier id must be `OracleScriptID`

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/client/proof.go [hunk 5]
```diff
@@ -430,7 +430,7 @@ func GetProofHandlerFn(cliCtx context.CLIContext) http.HandlerFunc {
 			return
 		}
 
-		key := zoracle.ResultStoreKey(int64(requestID), queryRequest.Request.OracleScriptID, queryRequest.Request.Calldata)
+		key := zoracle.ResultStoreKey(zoracle.RequestID(requestID), queryRequest.Request.OracleScriptID, queryRequest.Request.Calldata)
```
Comment (sorawit) [inline chain/client/proof.go]: This is not natural. Let's have the proper type when the variable is declared.

File: chain/cmd/bandoracled/main.go [hunk 6]
```diff
@@ -142,20 +142,20 @@ func handleRequest(requestID int64) (sdk.TxResponse, error) {
 		if info.err != nil {
 			return sdk.TxResponse{}, info.err
 		}
-		reports = append(reports, zoracle.NewRawDataReport(info.externalID, info.answer))
+		reports = append(reports, zoracle.NewRawDataReport(zoracle.ExternalID(info.externalID), info.answer))
```
Comment (sorawit) [inline chain/cmd/bandoracled/main.go]: smelly.

File: chain/cmd/bandoracled/main.go [hunk 7]
```diff
@@ -142,20 +142,20 @@ func handleRequest(requestID int64) (sdk.TxResponse, error) {
...
 	}
 
 	sort.Slice(reports, func(i, j int) bool {
 		return reports[i].ExternalDataID < reports[j].ExternalDataID
 	})
 
 	return bandClient.SendTransaction(
-		zoracle.NewMsgReportData(requestID, reports, sdk.ValAddress(bandClient.Sender())),
+		zoracle.NewMsgReportData(zoracle.RequestID(requestID), reports, sdk.ValAddress(bandClient.Sender())),
```
Comment (sorawit) [inline chain/cmd/bandoracled/main.go]: Why is casting here necessary?

File: chain/cmd/bandsv/main.go [hunk 8]
```diff
@@ -134,7 +134,7 @@ func handleRequestData(c *gin.Context) {
 
 	txr, err := bandClient.SendTransaction(
 		zoracle.NewMsgRequestData(
-			req.OracleScriptID,
+			zoracle.OracleScriptID(req.OracleScriptID),
```
Comment (sorawit) [inline chain/cmd/bandsv/main.go]: also here

File: chain/cmd/bandsv/main.go [hunk 9]
```diff
@@ -175,7 +175,7 @@ func handleRequestData(c *gin.Context) {
 	if reqType == Synchronous {
 		c.JSON(200, OracleRequestResp{
 			TxHash:    txr.TxHash,
-			RequestID: requestID,
+			RequestID: zoracle.RequestID(requestID),
```
Comment (sorawit) [inline chain/cmd/bandsv/main.go]: also here

File: chain/cmd/bandsv/main.go [hunk 10]
```diff
@@ -201,7 +201,7 @@ func handleRequestData(c *gin.Context) {
 
 			c.JSON(200, OracleRequestResp{
 				TxHash:    txr.TxHash,
-				RequestID: requestID,
+				RequestID: zoracle.RequestID(requestID),
```
Comment (sorawit) [inline chain/cmd/bandsv/main.go]: also here

File: chain/x/zoracle/client/cli/tx.go [hunk 11]
```diff
@@ -108,7 +108,7 @@ $ %s tx zoracle request 1 --calldata 1234abcdef --requested-validator-count 4 --
 			}
 
 			msg := types.NewMsgRequestData(
-				oracleScriptID,
+				types.OracleScriptID(oracleScriptID),
```
Comment (sorawit) [inline chain/x/zoracle/client/cli/tx.go]: Let's have the type when the variable is created.

File: chain/x/zoracle/client/cli/tx.go [hunk 12]
```diff
@@ -176,15 +176,16 @@ $ %s tx zoracle report 1 1:172.5 2:HELLOWORLD --from mykey
 				if err != nil {
 					return err
 				}
-				dataset = append(dataset, types.NewRawDataReport(externalID, []byte(reportRaw[1])))
+
+				dataset = append(dataset, types.NewRawDataReport(types.ExternalID(externalID), []byte(reportRaw[1])))
```
Comment (sorawit) [inline chain/x/zoracle/client/cli/tx.go]: also here

File: chain/x/zoracle/client/cli/tx.go [hunk 13]
```diff
@@ -176,15 +176,16 @@ $ %s tx zoracle report 1 1:172.5 2:HELLOWORLD --from mykey
...
 			}
 
 			// Sort data reports by external ID
 			sort.Slice(dataset, func(i, j int) bool {
 				return dataset[i].ExternalDataID < dataset[j].ExternalDataID
 			})
 
-			msg := types.NewMsgReportData(requestID, dataset, sdk.ValAddress(cliCtx.GetFromAddress()))
+			msg := types.NewMsgReportData(types.RequestID(requestID), dataset, sdk.ValAddress(cliCtx.GetFromAddress()))
```
Comment (sorawit) [inline chain/x/zoracle/client/cli/tx.go]: also here

File: chain/x/zoracle/internal/keeper/request.go [hunk 14]
```diff
@@ -140,10 +140,10 @@ func (k Keeper) ShouldBecomePendingResolve(ctx sdk.Context, id int64) bool {
 }
 
 // AddPendingRequest checks and append new request id to list if id already existed in list, it will return error.
-func (k Keeper) AddPendingRequest(ctx sdk.Context, requestID int64) sdk.Error {
+func (k Keeper) AddPendingRequest(ctx sdk.Context, requestID types.RequestID) sdk.Error {
 	pendingList := k.GetPendingResolveList(ctx)
 	for _, entry := range pendingList {
-		if requestID == entry {
+		if requestID == types.RequestID(entry) {
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/request.go]: Is this cast necessary?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/client/proof.go [hunk 15]
```diff
@@ -399,7 +399,8 @@ func MakeOtherStoresMerkleHash(mspo rootmulti.MultiStoreProofOp) (cmn.HexBytes,
 func GetProofHandlerFn(cliCtx context.CLIContext) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		vars := mux.Vars(r)
-		requestID, err := strconv.ParseUint(vars[requestIDTag], 10, 64)
+		intRequestID, err := strconv.ParseUint(vars[requestIDTag], 10, 64)
+		requestID := zoracle.RequestID(intRequestID)
```
Comment (sorawit) [inline chain/client/proof.go]: Check for errors first before proceeding. Make it a habit. Same for all places below.

File: chain/x/zoracle/internal/keeper/data_source.go [hunk 16]
```diff
@@ -26,12 +26,12 @@ func (k Keeper) AddDataSource(ctx sdk.Context, owner sdk.AccAddress, name string
 	}
 
 	newDataSource := types.NewDataSource(owner, name, fee, executable)
-	k.SetDataSource(ctx, newDataSourceID, newDataSource)
+	k.SetDataSource(ctx, types.DataSourceID(newDataSourceID), newDataSource)
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/data_source.go]: Is this necessary?

File: chain/x/zoracle/internal/keeper/keeper.go [hunk 17]
```diff
@@ -120,8 +120,8 @@ func (keeper Keeper) GetParams(ctx sdk.Context) types.Params {
 }
 
 // GetRequestCount returns the current number of all requests ever exist.
-func (k Keeper) GetRequestCount(ctx sdk.Context) int64 {
-	var requestNumber int64
+func (k Keeper) GetRequestCount(ctx sdk.Context) types.RequestID {
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/keeper.go]: This should return `int64`. It's request count, not request ID

File: chain/x/zoracle/internal/keeper/querier.go [hunk 18]
```diff
@@ -98,7 +98,7 @@ func queryDataSourceByID(ctx sdk.Context, path []string, req abci.RequestQuery,
 		return nil, sdk.ErrInternal(fmt.Sprintf("wrong format for data source id %s", err.Error()))
 	}
 
-	dataSource, sdkErr := keeper.GetDataSource(ctx, id)
+	dataSource, sdkErr := keeper.GetDataSource(ctx, types.DataSourceID(id))
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/querier.go]: Cast type when creating the variable.

File: chain/x/zoracle/internal/keeper/querier.go [hunk 19]
```diff
@@ -247,12 +247,12 @@ func queryRequests(
 
 	requests := make([]types.RequestQuerierInfo, 0)
 	allRequestsCount := keeper.GetRequestCount(ctx)
-	limit := startID + numberOfRequests - 1
+	limit := types.RequestID(startID + numberOfRequests - 1)
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/querier.go]: limit and and idx should be int!

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/client/cli/tx.go [hunk 20]
```diff
@@ -72,7 +72,8 @@ $ %s tx zoracle request 1 --calldata 1234abcdef --requested-validator-count 4 --
 
 			txBldr := auth.NewTxBuilderFromCLI().WithTxEncoder(utils.GetTxEncoder(cdc))
 
-			oracleScriptID, err := strconv.ParseInt(args[0], 10, 64)
+			int64OracleScriptID, err := strconv.ParseInt(args[0], 10, 64)
+			oracleScriptID := types.OracleScriptID(int64OracleScriptID)
```
Comment (sorawit) [inline chain/x/zoracle/client/cli/tx.go]: also fix error order here.

File: chain/x/zoracle/client/cli/tx.go [hunk 21]
```diff
@@ -161,7 +162,8 @@ $ %s tx zoracle report 1 1:172.5 2:HELLOWORLD --from mykey
 
 			txBldr := auth.NewTxBuilderFromCLI().WithTxEncoder(utils.GetTxEncoder(cdc))
 
-			requestID, err := strconv.ParseInt(args[0], 10, 64)
+			int64RequestID, err := strconv.ParseInt(args[0], 10, 64)
+			requestID := types.RequestID(int64RequestID)
```
Comment (sorawit) [inline chain/x/zoracle/client/cli/tx.go]: and here.

File: chain/x/zoracle/client/cli/tx.go [hunk 22]
```diff
@@ -172,10 +174,12 @@ $ %s tx zoracle report 1 1:172.5 2:HELLOWORLD --from mykey
 				if len(reportRaw) != 2 {
 					return fmt.Errorf("Invalid report format: %s", reportRaw[0])
 				}
-				externalID, err := strconv.ParseInt(reportRaw[0], 10, 64)
+				int64ExternalID, err := strconv.ParseInt(reportRaw[0], 10, 64)
+				externalID := types.ExternalID(int64ExternalID)
```
Comment (sorawit) [inline chain/x/zoracle/client/cli/tx.go]: and here

File: chain/x/zoracle/client/cli/tx.go [hunk 23]
```diff
@@ -325,7 +329,7 @@ $ %s tx zoracle edit-data-source 1 --name coingecko-price --script ../price.sh -
 			}
 
 			msg := types.NewMsgEditDataSource(
-				id,
+				types.DataSourceID(id),
```
Comment (sorawit) [inline chain/x/zoracle/client/cli/tx.go]: Let's make id have the correct type since its creation.

File: chain/x/zoracle/client/cli/tx.go [hunk 24]
```diff
@@ -457,7 +461,7 @@ $ %s tx zoracle edit-oracle-script 1 --name eth-price --script ../eth_price.wasm
 			}
 
 			msg := types.NewMsgEditOracleScript(
-				id,
+				types.OracleScriptID(id),
```
Comment (sorawit) [inline chain/x/zoracle/client/cli/tx.go]: Let's make id have the correct type since its creation.

**Review by sorawit — CHANGES_REQUESTED** [review]
Looks good overall. Just a few more comments!

File: chain/cmd/bandsv/main.go [hunk 25]
```diff
@@ -151,12 +151,13 @@ func handleRequestData(c *gin.Context) {
...
 
-	requestID := int64(0)
+	requestID := zoracle.RequestID(0)
 	for _, event := range txr.Events {
 		if event.Type == "request" {
 			for _, attr := range event.Attributes {
 				if string(attr.Key) == "id" {
-					requestID, err = strconv.ParseInt(attr.Value, 10, 64)
+					int64RequstID, err := strconv.ParseInt(attr.Value, 10, 64)
+					requestID = zoracle.RequestID(int64RequstID)
```
Comment (sorawit) [inline chain/cmd/bandsv/main.go]: Move error check up.

**Review by sorawit — APPROVED** [review]

**Review by sorawit — CHANGES_REQUESTED** [review]
please fix

**Review by sorawit — DISMISSED** [review]

**Review by sorawit — DISMISSED** [review]
Looks great. Thanks!

File: chain/client/proof.go [hunk 26]
```diff
@@ -418,7 +420,7 @@ func GetProofHandlerFn(cliCtx context.CLIContext) http.HandlerFunc {
 		}
 
 		var queryRequest zoracle.RequestQuerierInfo
-		res, _, err := cliCtx.QueryWithData(fmt.Sprintf("custom/zoracle/request/%d", requestID), nil)
+		res, _, err := cliCtx.QueryWithData(fmt.Sprintf("custom/zoracle/request/%d", uint64(requestID)), nil)
```
Comment (taobun) [inline chain/client/proof.go]: I think we don't need to cast in `Sprintf`

File: chain/cmd/bandoracled/main.go [hunk 27]
```diff
@@ -123,17 +123,17 @@ func handleRequest(requestID int64) (sdk.TxResponse, error) {
...
 	}
 
 	chanQueryParallelInfo := make(chan queryParallelInfo, len(request.RawDataRequests))
 	for _, rawRequest := range request.RawDataRequests {
-		go func(externalID, dataSourceID int64, calldata []byte) {
+		go func(externalID zoracle.ExternalID, dataSourceID zoracle.DataSourceID, calldata []byte) {
 			info := queryParallelInfo{externalID: externalID, answer: []byte{}, err: nil}
 			res, _, err := cliCtx.Query(
-				fmt.Sprintf("custom/zoracle/%s/%d", zoracle.QueryDataSourceByID, dataSourceID),
+				fmt.Sprintf("custom/zoracle/%s/%d", zoracle.QueryDataSourceByID, int64(dataSourceID)),
```
Comment (taobun) [inline chain/cmd/bandoracled/main.go]: ditto

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/handler_test.go [hunk 28]
```diff
@@ -594,7 +594,7 @@ func TestStopResolveWhenOutOfGas(t *testing.T) {
 		require.Equal(t, types.Success, actualRequest.ResolveStatus)
 	}
 
-	for i := int64(10); i <= 10; i++ {
+	for i := types.RequestID(10); i <= 10; i++ {
```
Comment (evilpeach) [inline chain/x/zoracle/handler_test.go]: What do you prefer? 

1. `for i := types.RequestID(1); i <= types.RequestID(10); i++ {...}` or 
2. `for i := types.RequestID(1); i <= 10; i++ {...}`

Please make it consistent with other test files.

**Review by evilpeach — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Review by evilpeach — APPROVED** [review]

**Pull request merged**